redshift/redshift.go

```diff
@@ -71,149 +67,296 @@ func (c sortableColumns) Less(i, j int) bool { return c[i].Ordinal < c[j].Ordina
+// of the db table and the last data in the table, if that exists
+// if the table does not exist it returns an empty table but does not error
+func (r *Redshift) GetTableMetadata(schema, tableName, dataDateCol string) (*Table, *time.Time, error) {
+	var cols []ColInfo
+
+	// does the table exist?
+	var placeholder string
+	q := fmt.Sprintf(existQueryFormat, schema, tableName)
+	if err := r.QueryRow(q).Scan(&placeholder); err != nil {
+		if err == sql.ErrNoRows {
+			log.Printf("schema: %s, table: %s does not exist", schema, tableName)
+			return nil, nil, sql.ErrNoRows
+		}
+		return nil, nil, fmt.Errorf("issue just checking if the table exists: %s", err)
 	}
-	gzipSQL := ""
-	if gzip {
-		gzipSQL = "GZIP"
+
+	// table exists, what are the columns?
+	rows, err := r.Query(fmt.Sprintf(schemaQueryFormat, schema, tableName))
+	if err != nil {
+		return nil, nil, fmt.Errorf("issue running column query: %s, err: %s", schemaQueryFormat, err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var c ColInfo
+		if err := rows.Scan(&c.Ordinal, &c.Name, &c.Type, &c.DefaultVal, &c.NotNull,
+			&c.PrimaryKey, &c.DistKey, &c.SortOrdinal,
+		); err != nil {
+			return nil, nil, fmt.Errorf("issue scanning column, err: %s", err)
+		}
+
+		cols = append(cols, c)
+	}
+	if err := rows.Err(); err != nil {
+		return nil, nil, fmt.Errorf("issue iterating over columns, err: %s", err)
+	}
+
+	// turn into Table struct
+	retTable := Table{
+		Name:    tableName,
+		Columns: cols,
+		Meta: Meta{
+			DataDateColumn: dataDateCol,
+			Schema:         schema,
+		},
+	}
+
+	// what's the last data in the table?
+	lastDataQuery := fmt.Sprintf("SELECT %s FROM %s.%s ORDER BY %s DESC LIMIT 1",
+		dataDateCol, schema, tableName, dataDateCol)
+	var lastData time.Time
+	if err = r.QueryRow(lastDataQuery).Scan(&lastData); err != nil {
+		return nil, nil, fmt.Errorf("issue running query: %s, err: %s", lastDataQuery, err)
+	}
+	return &retTable, &lastData, nil
+}
+
+func getColumnSQL(c ColInfo) string {
+	// note that we are relying on redshift to fail if we have created multiple sort keys
+	// currently we don't support that
+	defaultVal := ""
+	if c.DefaultVal != "" {
+		defaultVal = fmt.Sprintf("DEFAULT %s", c.DefaultVal)
+	}
+	notNull := ""
+	if c.NotNull {
+		notNull = "NOT NULL"
+	}
+	primaryKey := ""
+	if c.PrimaryKey {
+		primaryKey = "PRIMARY KEY"
+	}
+	sortKey := ""
+	if c.SortOrdinal == 1 {
+		sortKey = "SORTKEY"
+	}
+	distKey := ""
+	if c.DistKey {
+		distKey = "DISTKEY"
 	}
-	copySQL := `COPY "?"."?" FROM '?' WITH ? JSON '?' REGION '?' TIMEFORMAT 'auto' STATUPDATE ON COMPUPDATE ON %s`
-	copyStmt, err := r.Prepare(fmt.Sprintf(copySQL, credSQL))
+
+	return fmt.Sprintf(" %s %s %s %s %s %s %s", c.Name, typeMapping[c.Type], defaultVal, notNull, sortKey, primaryKey, distKey)
+}
+
+// RunCreateTable runs the full create table command in the provided transaction, given a
+// redshift representation of the table.
+func (r *Redshift) RunCreateTable(tx *sql.Tx, table Table) error {
```

Comment:
Purely a style opinion, but why is the name `RunCreateTable` instead of just `CreateTable`? IMHO, this is usually used when you have a fully prepared operation, like the result of `tx.Prepare` in this case.